source/slang/slang-check-decl.cpp

```diff
@@ -343,6 +343,11 @@ struct SemanticsDeclHeaderVisitor : public SemanticsDeclVisitorBase,
 
     void validateGenericConstraintSubType(GenericTypeConstraintDecl* decl, TypeExp type);
 
+    void checkForwardReferencesInGenericConstraint(GenericTypeConstraintDecl* decl);
+
+    // General utility function to collect all referenced declarations from a value
+    void collectReferencedDecls(Val* val, HashSet<Decl*>& outDecls);
```

Comment:
This is a general utility that can be useful outside this class. It is better to move it to `slang-check.h`.